File: milkdi-frontend/data/services/retailerService.js
import apiClient from "../api/apiClient";
import { API_ENDPOINTS } from "../api/apiEndPoint";

const retailerService = {
    getDashboardStats: async () => {
        console.trace("🕵️ DEBUG: retailerService.getDashboardStats called");
        const response = await apiClient.get(API_ENDPOINTS.RETAILER.DASHBOARD_STATS);
        return response.data;
    },

    getOrders: async (customerId = null, page = 1, limit = 10, statusFilter = 'All') => {
        const response = await apiClient.get(API_ENDPOINTS.RETAILER.ORDERS, {
            params: { customerId, page, limit, statusFilter }
        });
        
        // Handle the paginated object structure from backend
        const ordersRaw = response.data.data?.orders || response.data.orders || [];
        
        if (response.data.success && Array.isArray(ordersRaw)) {
            const mappedOrders = ordersRaw.map((order) => ({
                id: order.id || order.orderId,
                _id: order._id,
                date: order.date || new Date(order.createdAt).toLocaleDateString(),
                product: order.product || (order.items?.map(i => `${i.quantity}x ${i.product?.name || 'Product'}`).join(', ')) || 'No products',
                price: order.price || order.totalAmount || 0,
                status: order.status,
                payment: order.payment || order.paymentStatus,
                orderType: order.orderType,
                rider: order.rider,
                statusHistory: order.statusHistory || [],
                items: order.items || [],
                user: order.user || null,
                deliveryAddress: order.deliveryAddress || null,
                deliverySlot: order.deliverySlot || null
            }));
            
            if (response.data.data?.orders) {
                response.data.data.orders = mappedOrders;
            } else if (response.data.orders) {
                response.data.orders = mappedOrders;
            }
        }
        return response.data;
    },

    updateOrderStatus: async (orderId, status) => {
        const response = await apiClient.patch(API_ENDPOINTS.RETAILER.ORDER_STATUS, { orderId, status });
        return response.data;
    },

    getReviews: async () => {
        const response = await apiClient.get(API_ENDPOINTS.RETAILER.REVIEWS);
        return response.data;
    },

    getRevenueStats: async (range = 'month', startDate = null, endDate = null) => {
        const response = await apiClient.get(API_ENDPOINTS.RETAILER.REVENUE_STATS, { 
            params: { range, startDate, endDate } 
        });
        return response.data;
    },

    getCustomers: async () => {
        const response = await apiClient.get(API_ENDPOINTS.RETAILER.CUSTOMERS);
        return response.data;
    },

    getCategories: async () => {
        const response = await apiClient.get(API_ENDPOINTS.RETAILER.CATEGORIES);
        return response.data;
    },

    // Products
    getProducts: async () => {
        const response = await apiClient.get(API_ENDPOINTS.RETAILER.PRODUCTS);
        return response.data;
    },

    getProduct: async (id) => {
        const response = await apiClient.get(`${API_ENDPOINTS.RETAILER.PRODUCTS}/${id}`);
        return response.data;
    },

    createProduct: async (productData) => {
        const response = await apiClient.post(API_ENDPOINTS.RETAILER.PRODUCTS, productData);
        return response.data;
    },

    updateProduct: async (id, productData) => {
        const response = await apiClient.put(`${API_ENDPOINTS.RETAILER.PRODUCTS}/${id}`, productData);
        return response.data;
    },

    deleteProduct: async (id) => {
        const response = await apiClient.delete(`${API_ENDPOINTS.RETAILER.PRODUCTS}/${id}`);
        return response.data;
    },

    // Profile & Settings
    getProfile: async (userId) => {
        const response = await apiClient.get(`${API_ENDPOINTS.AUTH.ME}/${userId}`);
        return response.data;
    },

    updateProfile: async (profileData) => {
        const response = await apiClient.put(API_ENDPOINTS.RETAILER.PROFILE, profileData);
        return response.data;
    },

    // Payouts
    requestPayout: async (payoutData) => {
        const response = await apiClient.post(API_ENDPOINTS.PAYOUT.REQUEST, payoutData);
        return response.data;
    },

    getPayoutHistory: async (page = 1, limit = 10) => {
        const response = await apiClient.get(API_ENDPOINTS.PAYOUT.MY_HISTORY, {
            params: { page, limit }
        });
        return response.data;
    },

    // Shop Status
    toggleShopStatus: async () => {
        const response = await apiClient.patch(API_ENDPOINTS.RETAILER.TOGGLE_STATUS);
        return response.data;
    },

    getPrepList: async (date) => {
        const response = await apiClient.get(API_ENDPOINTS.RETAILER.PREP_LIST, { params: { date } });
        return response.data;
    },

    uploadImage: async (file) => {
        const formData = new FormData();
        formData.append("file", file);
        const response = await apiClient.post(API_ENDPOINTS.UPLOAD.IMAGE, formData, {
            headers: {
                "Content-Type": "multipart/form-data"
            }
        });
        return response.data;
    },

    // Riders
    getRiders: async () => {
        const response = await apiClient.get(API_ENDPOINTS.RIDER.RETAILER);
        return response.data;
    },

    addRider: async (riderData) => {
        const response = await apiClient.post(API_ENDPOINTS.RIDER.ADD, riderData);
        return response.data;
    },

    updateRiderStatus: async (riderId, status) => {
        const response = await apiClient.patch(`${API_ENDPOINTS.RIDER.RETAILER}/${riderId}/status`, { status });
        return response.data;
    },

    assignRider: async (orderId, riderId) => {
        const response = await apiClient.post(API_ENDPOINTS.RETAILER.ASSIGN_RIDER, { orderId, riderId });
        return response.data;
    },

    updateRider: async (riderId, riderData) => {
        const response = await apiClient.patch(`${API_ENDPOINTS.RIDER.RETAILER}/${riderId}`, riderData);
        return response.data;
    },

    deleteRider: async (riderId) => {
        const response = await apiClient.delete(`${API_ENDPOINTS.RIDER.RETAILER}/${riderId}`);
        return response.data;
    },

    addManualCustomer: async (customerData) => {
        const response = await apiClient.post(API_ENDPOINTS.RETAILER.CUSTOMERS, customerData);
        return response.data;
    },

    createManualOrder: async (orderData) => {
        const response = await apiClient.post(API_ENDPOINTS.RETAILER.ORDERS_MANUAL, orderData);
        return response.data;
    },

    settleCustomerDue: async (customerId, amount) => {
        const response = await apiClient.post(API_ENDPOINTS.RETAILER.CUSTOMERS_SETTLE_DUE, { customerId, amount });
        return response.data;
    },

    createManualSubscription: async (subscriptionData) => {
        const response = await apiClient.post(API_ENDPOINTS.RETAILER.SUBSCRIPTIONS_MANUAL, subscriptionData);
        return response.data;
    },

    getRetailerSubscriptions: async (customerId) => {
        const baseUrl = API_ENDPOINTS.RETAILER.SUBSCRIPTIONS;
        const url = customerId ? `${baseUrl}?customerId=${customerId}` : baseUrl;
        const response = await apiClient.get(url);
        return response.data;
    },
    
    searchAnything: async (query) => {
        const response = await apiClient.get(API_ENDPOINTS.RETAILER.SEARCH, { params: { q: query } });
        return response.data;
    },

    getDueOrdersForCustomer: async (customerId, type = 'due') => {
        const response = await apiClient.get(`${API_ENDPOINTS.RETAILER.CUSTOMERS}/${customerId}/due-orders`, { params: { type } });
        return response.data;
    },

    bulkProcessOrders: async () => {
        const response = await apiClient.post(API_ENDPOINTS.RETAILER.BULK_PROCESS_ORDERS);
        return response.data;
    },

    getBanks: async () => {
        const response = await apiClient.get(API_ENDPOINTS.RETAILER.BANKS);
        return response.data;
    },

    addBank: async (bankData) => {
        const response = await apiClient.post(API_ENDPOINTS.RETAILER.BANKS, bankData);
        return response.data;
    },

    deleteBank: async (id) => {
        const response = await apiClient.delete(`${API_ENDPOINTS.RETAILER.BANKS}/${id}`);
        return response.data;
    },

    setDefaultBank: async (id) => {
        const response = await apiClient.patch(`${API_ENDPOINTS.RETAILER.BANKS}/${id}/default`);
        return response.data;
    },

    // Delivery Charges (only for permitted retailers)
    getDeliveryCharges: async () => {
        const response = await apiClient.get("/delivery-charge/retailer-charges");
        return response.data;
    },

    updateDeliveryCharges: async (slabs, retailerMaxDeliveryKm) => {
        const response = await apiClient.put("/delivery-charge/retailer-charges", { slabs, retailerMaxDeliveryKm });
        return response.data;
    },

    getDeliveryIncome: async () => {
        const response = await apiClient.get("/delivery-charge/retailer-income");
        return response.data;
    },

    reverseGeocode: async (lat, lng) => {
        const response = await apiClient.get("/retailer/reverse-geocode", { params: { lat, lng } });
        return response.data;
    }
};

export default retailerService;

File: milkdi-continue-vendor-app/lib/app/data/services/retailer_service.dart
import '../network/api_client.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

class RetailerService {
  final ApiClient _apiClient;

  RetailerService(this._apiClient);

  Future<Map<String, dynamic>> getDashboardStats() async {
    try {
      final response = await _apiClient.get('/retailer/dashboard-stats');
      if (response != null && response['success'] == true) {
        return response['data'];
      }
      return {};
    } catch (e) {
      return {};
    }
  }

  Future<Map<String, dynamic>> getOrders() async {
    try {
      final response = await _apiClient.get('/retailer/orders');
      if (response != null && response['success'] == true) {
        return response['data'];
      }
      return {'orders': [], 'stats': {}};
    } catch (e) {
      return {'orders': [], 'stats': {}};
    }
  }

  Future<bool> updateOrderStatus(String orderId, String status) async {
    try {
      final response = await _apiClient.patch('/retailer/order-status', data: {
        'orderId': orderId,
        'status': status,
      });
      return response != null && response['success'] == true;
    } catch (e) {
      return false;
    }
  }

  Future<bool> toggleShopStatus() async {
    try {
      final response = await _apiClient.patch('/retailer/toggle-status');
      if (response != null && response['success'] == true) {
        return response['isShopActive'];
      }
      return false;
    } catch (e) {
      return false;
    }
  }

  // --- Product Management ---
  Future<List<dynamic>> getProducts() async {
    try {
      final response = await _apiClient.get('/retailer/products');
      if (response != null && response['success'] == true) {
        return response['data'];
      }
      return [];
    } catch (e) {
      return [];
    }
  }

  Future<List<dynamic>> getCategories() async {
    try {
      final response = await _apiClient.get('/retailer/categories');
      if (response != null && response['success'] == true) {
        return response['data'];
      }
      return [];
    } catch (e) {
      return [];
    }
  }

  Future<bool> createProduct(Map<String, dynamic> data) async {
    try {
      final response = await _apiClient.post('/retailer/products', data: data);
      return response != null && response['success'] == true;
    } catch (e) {
      return false;
    }
  }

  Future<bool> updateProduct(String id, Map<String, dynamic> data) async {
    try {
      final response = await _apiClient.put('/retailer/products/$id', data: data);
      return response != null && response['success'] == true;
    } catch (e) {
      return false;
    }
  }

  Future<bool> deleteProduct(String id) async {
    try {
      final response = await _apiClient.delete('/retailer/products/$id');
      return response != null && response['success'] == true;
    } catch (e) {
      return false;
    }
  }

  // --- Customer & CRM ---
  Future<List<dynamic>> getCustomers() async {
    try {
      final response = await _apiClient.get('/retailer/customers');
      if (response != null && response['success'] == true) {
        return response['data']['customers'] ?? [];
      }
      return [];
    } catch (e) {
      return [];
    }
  }

  Future<bool> addManualCustomer(Map<String, dynamic> data) async {
    try {
      final response = await _apiClient.post('/retailer/customers', data: data);
      return response != null && response['success'] == true;
    } catch (e) {
      return false;
    }
  }

  // --- Financials & Sales ---
  Future<Map<String, dynamic>> getRevenueStats() async {
    try {
      final response = await _apiClient.get('/retailer/revenue-stats');
      if (response != null && response['success'] == true) {
        return response['data'];
      }
      return {};
    } catch (e) {
      return {};
    }
  }

  // --- Customer Detailed Support ---
  Future<Map<String, dynamic>> getCustomerOrders(String customerId) async {
    try {
      final response = await _apiClient.get('/retailer/orders', queryParameters: {'customerId': customerId});
      if (response != null && response['success'] == true) {
        return response['data'];
      }
      return {'orders': []};
    } catch (e) {
      return {'orders': []};
    }
  }

  Future<Map<String, dynamic>> getDueOrdersForCustomer(String customerId) async {
    try {
      final response = await _apiClient.get('/retailer/customers/$customerId/due-orders');
      if (response != null && response['success'] == true) {
        return response['data'];
      }
      return {};
    } catch (e) {
      return {};
    }
  }

  Future<bool> settleCustomerDue(String customerId, double amount) async {
    try {
      final response = await _apiClient.post('/retailer/customers/settle-due', data: {
        'customerId': customerId,
        'amount': amount,
      });
      return response != null && response['success'] == true;
    } catch (e) {
      return false;
    }
  }

  Future<bool> createManualOrder(Map<String, dynamic> data) async {
    try {
      final response = await _apiClient.post('/retailer/orders/manual', data: data);
      return response != null && response['success'] == true;
    } catch (e) {
      return false;
    }
  }

  Future<bool> createManualSubscription(Map<String, dynamic> data) async {
    try {
      final response = await _apiClient.post('/retailer/subscriptions/manual', data: data);
      return response != null && response['success'] == true;
    } catch (e) {
      return false;
    }
  }

  Future<List<dynamic>> getRiders() async {
    try {
      final response = await _apiClient.get('/rider/retailer');
      if (response != null && response['success'] == true) {
        return response['data'];
      }
      return [];
    } catch (e) {
      return [];
    }
  }

  Future<bool> assignRider(String orderId, String riderId) async {
    try {
      final response = await _apiClient.post('/retailer/assign-rider', data: {
        'orderId': orderId,
        'riderId': riderId,
      });
      return response != null && response['success'] == true;
    } catch (e) {
      return false;
    }
  }
}

final retailerServiceProvider = Provider<RetailerService>((ref) {
  final apiClient = ref.watch(apiClientProvider);
  return RetailerService(apiClient);
});
